### ISSUE TITLE
FR: new onceWithLatestFrom to use streams' last values once in imperative scenarios

### ISSUE THREAD

**Issue by moniuch** [ReactiveX/rxjs#3185]
Feature Request

**RxJS version:**
5.5.x

In Angular 5 I often have to mix imperative code with reactive code, such as - when a method on a container is expected to trigger an action, and the subject of the action are objects taken from ngrx store. An example:

`<button (click)="fetchData()">Reload</button>`

```javascript
fetchData() {
    of(null).pipe(
      withLatestFrom(this.store.select(fromRoot.getSearchParams)),
      tap(([_, searchParams]) => this.store.dispatch(new actions.LoadData(searchParams))),
      take(1),
    )
    .subscribe();
  }
```

What is peculiar in the above code is the usage of: `of(null)` (as a way to enable `withLatestFrom`) combined with `take(1)` as a way to grab current value of an observable without waiting for its next emission, and perform the action once.

I was thinking, couldn't we perhaps have a special operator that would allow us to simplify this kind of imperative scenarios, and say something like:

```javascript
fetchData() {
  onceWithLatestFrom(
    this.store.select(fromRoot.getSearchParams),
    // optionally more observables here and a callback
  ).pipe(
    tap(([searchParams]) => this.store.dispatch(new actions.LoadData(searchParams)))
  )
  .subscribe();
}
```

I will not be surprised if I hear from you that this code would violate the reactive principle of the library and it should not be extended only to provide workarounds. However I would see a benefit in how much cleaner my code would get.

On a side note: I wish there was a way in Angular to bind events into a `.fromEvent` observable - this would allow to easily combine with other streams and avoid mixing the techniques. This is however still undecided how and when Angular team will provide it:
https://github.com/angular/angular/issues/13248

**Comment by staltz** [comment]
1. Make the store.select output observable become a BehaviorSubject, which has this "current value" functionality built-in
2. Move that code that is now in `tap` and put it in subscribe. A subscribe with empty body is usually a bad sign

**Comment by moniuch** [comment]
Thanks @staltz:
1. I have been avoiding putting a BehaviorSubject on top of a store which is already a subject by itself - considered it some kind of overcomplication.
2. How can having an action dispatched from within `tap` be functionally different from dispatching an action from within the subscribe body?

**Comment by staltz** [comment]
1. What's your concern with BehaviorSubject becoming overcomplication?
2. Not much, but it's a bad habit to do important side effects in `tap`. For instance, if you forget to subscribe, then the tap side effect won't happen.

**Comment by moniuch** [comment]
1. Perhaps redundancy would be a better word: if I am on top of a BehaviorSubject, why would I need yet another subject to achieve my goal - that way of reasoning. Now you specified why, but I still think we could have an easier way to do that. Something is standing in the way here IMO...
2. Well, at the begining of using Observables, true, I did often forget to subscribe. But this is a basic lesson to learn. Also, I liked that pattern, where - on a container level, I first declared and defined my observables, and then I gathered all the subscriptions in one block (so to unsubscribe them en block in ngOnDestroy):

```
this.subscriptions = [
  myObs1$.subscribe(),
  myObs2$.subscribe(),
  // etc.
]
```
